subprojects/core/src/main/java/org/gradle/initialization/properties/DefaultProjectPropertiesLoader.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright 2023 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.initialization.properties;
+
+import org.gradle.api.internal.StartParameterInternal;
+import org.gradle.initialization.DefaultGradleProperties;
+import org.gradle.initialization.Environment;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class DefaultProjectPropertiesLoader implements ProjectPropertiesLoader {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(DefaultProjectPropertiesLoader.class);
+
+    private final Environment environment;
+
+    private final StartParameterInternal startParameter;
+
+    public DefaultProjectPropertiesLoader(StartParameterInternal startParameter, Environment environment) {
+        this.environment = environment;
+        this.startParameter = startParameter;
+    }
+
+    @Override
+    public GradlePropertiesInternal loadProjectProperties() {
```

Comment:
We're populating the `overrideProperties` only, maybe we can return only these as a `Map` instead? Then, to simplify things further:
- Remove `getDefaultProperties` and `getOverrideProperties` from `GradlePropertiesInternal`
- add `updateDefaultProperties` to `GradlePropertiesInternal` and rename it to `MutableGradleProperties`. This also addresses `GradleProperties` already being in the `internal` package, so `GradlePropertiesInternal` is very internal.
- use the new method in `loadGradlePropertiesFrom`, but only expose the result as `GradleProperties`. 
I don't think we need a truly immutable implementation, though.